<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>9 Data Visualization with ggplot2 | Introduction to R for Biologists</title>
  <meta name="description" content="This tutorial covers workshop material" />
  <meta name="generator" content="bookdown 0.17 and GitBook 2.6.7" />

  <meta property="og:title" content="9 Data Visualization with ggplot2 | Introduction to R for Biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This tutorial covers workshop material" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="9 Data Visualization with ggplot2 | Introduction to R for Biologists" />
  
  <meta name="twitter:description" content="This tutorial covers workshop material" />
  

<meta name="author" content="Alana Schick" />


<meta name="date" content="2020-03-13" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="manipulating-data-in-the-tidyverse.html"/>
<link rel="next" href="project.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Intro R Workshop</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Welcome</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#workshop-schedule"><i class="fa fa-check"></i><b>1.1</b> Workshop Schedule</a><ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#day-1"><i class="fa fa-check"></i>Day 1</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#day-2"><i class="fa fa-check"></i>Day 2</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#important-links"><i class="fa fa-check"></i><b>1.2</b> Important links</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="pre-workshop.html"><a href="pre-workshop.html"><i class="fa fa-check"></i><b>2</b> Pre-workshop</a><ul>
<li class="chapter" data-level="2.1" data-path="pre-workshop.html"><a href="pre-workshop.html#installation"><i class="fa fa-check"></i><b>2.1</b> Installation</a><ul>
<li class="chapter" data-level="2.1.1" data-path="pre-workshop.html"><a href="pre-workshop.html#windows"><i class="fa fa-check"></i><b>2.1.1</b> Windows</a></li>
<li class="chapter" data-level="2.1.2" data-path="pre-workshop.html"><a href="pre-workshop.html#macos"><i class="fa fa-check"></i><b>2.1.2</b> MacOS</a></li>
<li class="chapter" data-level="2.1.3" data-path="pre-workshop.html"><a href="pre-workshop.html#linux"><i class="fa fa-check"></i><b>2.1.3</b> Linux</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="pre-workshop.html"><a href="pre-workshop.html#datacamp-courses"><i class="fa fa-check"></i><b>2.2</b> Datacamp Courses</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>3</b> Introduction</a><ul>
<li class="chapter" data-level="3.1" data-path="introduction.html"><a href="introduction.html#r-and-the-rstudio-ide"><i class="fa fa-check"></i><b>3.1</b> R and the RStudio IDE</a></li>
<li class="chapter" data-level="3.2" data-path="introduction.html"><a href="introduction.html#why-learn-r"><i class="fa fa-check"></i><b>3.2</b> Why learn R</a><ul>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#r-does-not-involve-a-lot-of-pointing-and-clicking---thats-a-good-thing"><i class="fa fa-check"></i>R does not involve a lot of pointing and clicking - that’s a good thing!</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#r-code-is-great-for-reproducibility"><i class="fa fa-check"></i>R code is great for reproducibility</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#r-is-interdisciplinary-and-extensible"><i class="fa fa-check"></i>R is interdisciplinary and extensible</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#r-works-on-data-of-all-shapes-and-sizes"><i class="fa fa-check"></i>R works on data of all shapes and sizes</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#r-produces-high-quality-graphics"><i class="fa fa-check"></i>R produces high-quality graphics</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#r-has-a-large-and-welcoming-community"><i class="fa fa-check"></i>R has a large and welcoming community</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#interacting-with-r"><i class="fa fa-check"></i>Interacting with R</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="introduction.html"><a href="introduction.html#creating-an-rstudio-project"><i class="fa fa-check"></i><b>3.3</b> Creating an RStudio project</a></li>
<li class="chapter" data-level="3.4" data-path="introduction.html"><a href="introduction.html#creating-your-first-r-script"><i class="fa fa-check"></i><b>3.4</b> Creating your first R script</a><ul>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#organizing-your-working-directory"><i class="fa fa-check"></i>Organizing your working directory</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="introduction.html"><a href="introduction.html#how-to-learn-more"><i class="fa fa-check"></i><b>3.5</b> How to learn more?</a><ul>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#seeking-help"><i class="fa fa-check"></i>Seeking help</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#i-am-stuck-i-get-an-error-message-that-i-dont-understand"><i class="fa fa-check"></i>I am stuck… I get an error message that I don’t understand</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#asking-for-help"><i class="fa fa-check"></i>Asking for help</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#where-to-ask-for-help"><i class="fa fa-check"></i>Where to ask for help?</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#how-to-ask-for-r-help-useful-guidelines"><i class="fa fa-check"></i>How to ask for R help useful guidelines</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="installing-r-packages.html"><a href="installing-r-packages.html"><i class="fa fa-check"></i><b>4</b> Installing R Packages</a><ul>
<li class="chapter" data-level="4.1" data-path="installing-r-packages.html"><a href="installing-r-packages.html#what-are-r-packages"><i class="fa fa-check"></i><b>4.1</b> What are R packages?</a></li>
<li class="chapter" data-level="4.2" data-path="installing-r-packages.html"><a href="installing-r-packages.html#installing-r-packages-from-different-sources"><i class="fa fa-check"></i><b>4.2</b> Installing R packages from different sources</a><ul>
<li class="chapter" data-level="" data-path="installing-r-packages.html"><a href="installing-r-packages.html#cran"><i class="fa fa-check"></i>1. CRAN</a></li>
<li class="chapter" data-level="" data-path="installing-r-packages.html"><a href="installing-r-packages.html#bioconductor"><i class="fa fa-check"></i>2. Bioconductor</a></li>
<li class="chapter" data-level="" data-path="installing-r-packages.html"><a href="installing-r-packages.html#github"><i class="fa fa-check"></i>3. Github</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="installing-r-packages.html"><a href="installing-r-packages.html#loading-a-package"><i class="fa fa-check"></i><b>4.3</b> Loading a package</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="r-basics.html"><a href="r-basics.html"><i class="fa fa-check"></i><b>5</b> R Basics</a><ul>
<li class="chapter" data-level="5.1" data-path="r-basics.html"><a href="r-basics.html#creating-objects-in-r"><i class="fa fa-check"></i><b>5.1</b> Creating objects in R</a></li>
<li class="chapter" data-level="5.2" data-path="r-basics.html"><a href="r-basics.html#objects-vs.variables"><i class="fa fa-check"></i><b>5.2</b> Objects vs. variables</a></li>
<li class="chapter" data-level="5.3" data-path="r-basics.html"><a href="r-basics.html#comments"><i class="fa fa-check"></i><b>5.3</b> Comments</a></li>
<li class="chapter" data-level="5.4" data-path="r-basics.html"><a href="r-basics.html#functions-and-their-arguments"><i class="fa fa-check"></i><b>5.4</b> Functions and their arguments</a></li>
<li class="chapter" data-level="5.5" data-path="r-basics.html"><a href="r-basics.html#vectors"><i class="fa fa-check"></i><b>5.5</b> Vectors</a></li>
<li class="chapter" data-level="5.6" data-path="r-basics.html"><a href="r-basics.html#indexing-vectors"><i class="fa fa-check"></i><b>5.6</b> Indexing vectors</a><ul>
<li class="chapter" data-level="" data-path="r-basics.html"><a href="r-basics.html#adding-values-to-vectors"><i class="fa fa-check"></i>Adding values to vectors</a></li>
<li class="chapter" data-level="" data-path="r-basics.html"><a href="r-basics.html#replacing-values-in-existing-vectors"><i class="fa fa-check"></i>Replacing values in existing vectors</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="r-basics.html"><a href="r-basics.html#conditional-subsetting"><i class="fa fa-check"></i><b>5.7</b> Conditional subsetting</a></li>
<li class="chapter" data-level="5.8" data-path="r-basics.html"><a href="r-basics.html#so-we-can-use-this-to-select-only-the-values-above-50"><i class="fa fa-check"></i><b>5.8</b> so we can use this to select only the values above 50</a></li>
<li class="chapter" data-level="5.9" data-path="r-basics.html"><a href="r-basics.html#missing-data"><i class="fa fa-check"></i><b>5.9</b> Missing data</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="starting-with-data.html"><a href="starting-with-data.html"><i class="fa fa-check"></i><b>6</b> Starting with Data</a><ul>
<li class="chapter" data-level="6.1" data-path="starting-with-data.html"><a href="starting-with-data.html#download-the-data"><i class="fa fa-check"></i><b>6.1</b> Download the data</a></li>
<li class="chapter" data-level="6.2" data-path="starting-with-data.html"><a href="starting-with-data.html#csv-files"><i class="fa fa-check"></i><b>6.2</b> CSV files</a></li>
<li class="chapter" data-level="6.3" data-path="starting-with-data.html"><a href="starting-with-data.html#sidenote-importing-data-from-excel"><i class="fa fa-check"></i><b>6.3</b> Sidenote: Importing data from Excel</a></li>
<li class="chapter" data-level="6.4" data-path="starting-with-data.html"><a href="starting-with-data.html#import-data"><i class="fa fa-check"></i><b>6.4</b> Import data</a></li>
<li class="chapter" data-level="6.5" data-path="starting-with-data.html"><a href="starting-with-data.html#what-are-data-frames"><i class="fa fa-check"></i><b>6.5</b> What are data frames?</a><ul>
<li class="chapter" data-level="" data-path="starting-with-data.html"><a href="starting-with-data.html#saving-or-writing-a-data-frame"><i class="fa fa-check"></i>Saving or “writing” a data frame</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="starting-with-data.html"><a href="starting-with-data.html#what-are-factors"><i class="fa fa-check"></i><b>6.6</b> What are factors?</a><ul>
<li class="chapter" data-level="" data-path="starting-with-data.html"><a href="starting-with-data.html#the-sign"><i class="fa fa-check"></i>The $ sign</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="starting-with-data.html"><a href="starting-with-data.html#plotting-and-ordering-factors"><i class="fa fa-check"></i><b>6.7</b> Plotting and ordering factors</a></li>
<li class="chapter" data-level="6.8" data-path="starting-with-data.html"><a href="starting-with-data.html#subsetting-data-frames"><i class="fa fa-check"></i><b>6.8</b> Subsetting data frames</a></li>
<li class="chapter" data-level="6.9" data-path="starting-with-data.html"><a href="starting-with-data.html#other-data-types"><i class="fa fa-check"></i><b>6.9</b> Other data types</a><ul>
<li class="chapter" data-level="" data-path="starting-with-data.html"><a href="starting-with-data.html#matrices"><i class="fa fa-check"></i>Matrices</a></li>
<li class="chapter" data-level="" data-path="starting-with-data.html"><a href="starting-with-data.html#lists"><i class="fa fa-check"></i>Lists</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="conditional-statements-and-loops.html"><a href="conditional-statements-and-loops.html"><i class="fa fa-check"></i><b>7</b> Conditional Statements and Loops</a><ul>
<li class="chapter" data-level="7.1" data-path="conditional-statements-and-loops.html"><a href="conditional-statements-and-loops.html#conditional-statements"><i class="fa fa-check"></i><b>7.1</b> Conditional statements</a><ul>
<li class="chapter" data-level="" data-path="conditional-statements-and-loops.html"><a href="conditional-statements-and-loops.html#if"><i class="fa fa-check"></i>if</a></li>
<li class="chapter" data-level="" data-path="conditional-statements-and-loops.html"><a href="conditional-statements-and-loops.html#ifelse"><i class="fa fa-check"></i>if…else</a></li>
<li class="chapter" data-level="" data-path="conditional-statements-and-loops.html"><a href="conditional-statements-and-loops.html#nested-ifelse-statements"><i class="fa fa-check"></i>Nested if…else statements</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="conditional-statements-and-loops.html"><a href="conditional-statements-and-loops.html#loops"><i class="fa fa-check"></i><b>7.2</b> Loops</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html"><i class="fa fa-check"></i><b>8</b> Manipulating Data in the tidyverse</a><ul>
<li class="chapter" data-level="8.1" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html#what-is-the-tidyverse"><i class="fa fa-check"></i><b>8.1</b> What is the Tidyverse</a></li>
<li class="chapter" data-level="8.2" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html#tibbles"><i class="fa fa-check"></i><b>8.2</b> Tibbles</a><ul>
<li class="chapter" data-level="" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html#create-a-tibble"><i class="fa fa-check"></i>Create a tibble</a></li>
<li class="chapter" data-level="" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html#reading-in-tibbles"><i class="fa fa-check"></i>Reading in tibbles</a></li>
<li class="chapter" data-level="" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html#data-frames-versus-tibbles"><i class="fa fa-check"></i>Data frames versus tibbles</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html#exploring-tibbles"><i class="fa fa-check"></i><b>8.3</b> Exploring tibbles</a></li>
<li class="chapter" data-level="8.4" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html#tidy-data"><i class="fa fa-check"></i><b>8.4</b> Tidy data</a><ul>
<li><a href="manipulating-data-in-the-tidyverse.html#gather"><code>gather()</code></a></li>
<li><a href="manipulating-data-in-the-tidyverse.html#separate"><code>separate()</code></a></li>
<li><a href="manipulating-data-in-the-tidyverse.html#spread"><code>spread()</code></a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html#what-is-dplyr"><i class="fa fa-check"></i><b>8.5</b> What is dplyr?</a></li>
<li class="chapter" data-level="8.6" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html#selecting-columns-and-filtering-rows"><i class="fa fa-check"></i><b>8.6</b> Selecting columns and filtering rows</a></li>
<li class="chapter" data-level="8.7" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html#pipes"><i class="fa fa-check"></i><b>8.7</b> Pipes</a></li>
<li class="chapter" data-level="8.8" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html#mutate"><i class="fa fa-check"></i><b>8.8</b> Mutate</a></li>
<li class="chapter" data-level="8.9" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html#split-apply-combine-data-analysis-and-the-summarize-function"><i class="fa fa-check"></i><b>8.9</b> Split-apply-combine data analysis and the summarize() function</a></li>
<li class="chapter" data-level="8.10" data-path="manipulating-data-in-the-tidyverse.html"><a href="manipulating-data-in-the-tidyverse.html#other-great-resources"><i class="fa fa-check"></i><b>8.10</b> Other great resources</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html"><i class="fa fa-check"></i><b>9</b> Data Visualization with ggplot2</a><ul>
<li class="chapter" data-level="9.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#data"><i class="fa fa-check"></i><b>9.1</b> Data</a></li>
<li class="chapter" data-level="9.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#plotting-with-ggplot2"><i class="fa fa-check"></i><b>9.2</b> Plotting with ggplot2</a><ul>
<li class="chapter" data-level="" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#principles-of-effective-display"><i class="fa fa-check"></i>Principles of effective display</a></li>
<li class="chapter" data-level="" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#building-plots-iteratively"><i class="fa fa-check"></i>Building plots iteratively</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#faceting"><i class="fa fa-check"></i><b>9.3</b> Faceting</a></li>
<li class="chapter" data-level="9.4" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#barplots"><i class="fa fa-check"></i><b>9.4</b> Barplots</a></li>
<li class="chapter" data-level="9.5" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#histograms"><i class="fa fa-check"></i><b>9.5</b> Histograms</a></li>
<li class="chapter" data-level="9.6" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#boxplots"><i class="fa fa-check"></i><b>9.6</b> Boxplots</a><ul>
<li class="chapter" data-level="" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#the-five-number-summary"><i class="fa fa-check"></i>The Five Number Summary</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#themes"><i class="fa fa-check"></i><b>9.7</b> Themes</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="project.html"><a href="project.html"><i class="fa fa-check"></i><b>10</b> Project</a><ul>
<li class="chapter" data-level="" data-path="project.html"><a href="project.html#introduction-1"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="project.html"><a href="project.html#task-0-download-the-data"><i class="fa fa-check"></i>Task 0: Download the data</a></li>
<li class="chapter" data-level="" data-path="project.html"><a href="project.html#task-1-data-import-and-examination"><i class="fa fa-check"></i>Task 1: Data import and examination</a><ul>
<li class="chapter" data-level="" data-path="project.html"><a href="project.html#task-1-solution"><i class="fa fa-check"></i>Task 1 Solution</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="project.html"><a href="project.html#task-2-data-tidying"><i class="fa fa-check"></i>Task 2: Data Tidying</a><ul>
<li class="chapter" data-level="" data-path="project.html"><a href="project.html#task-2-solution"><i class="fa fa-check"></i>Task 2 Solution</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="project.html"><a href="project.html#task-3-plots"><i class="fa fa-check"></i>Task 3: Plots</a><ul>
<li class="chapter" data-level="" data-path="project.html"><a href="project.html#task-3-solution"><i class="fa fa-check"></i>Task 3 Solution</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="project.html"><a href="project.html#task-4-heatmap"><i class="fa fa-check"></i>Task 4: Heatmap</a><ul>
<li class="chapter" data-level="" data-path="project.html"><a href="project.html#task-4-solution"><i class="fa fa-check"></i>Task 4 Solution</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="summary.html"><a href="summary.html"><i class="fa fa-check"></i><b>11</b> Summary</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to R for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-visualization-with-ggplot2" class="section level1">
<h1><span class="header-section-number">9</span> Data Visualization with ggplot2</h1>
<div id="data" class="section level2">
<h2><span class="header-section-number">9.1</span> Data</h2>
<p>This section uses a different data set from the same experiment - the LTEE (long-term evolution experiment). This data was pulished in: Tempo and mode of genome evolution in a 50,000-generation experiment <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/">Tenaillon et al 2016</a>.</p>
<p>Download the data onto your computer from <a href="https://www.dropbox.com/s/1z7qrbke9l6o6ee/combined_tidy_vcf.csv?dl=0">this</a> dropbox link and move it into the <code>data</code> directory of your RStudio project.</p>
<div class="box">
<h4 id="exercise-21" class="unnumbered">EXERCISE</h4>
<p>Read the data into R. Call this data <code>variants</code>. Ensure that you have read the data in by calling <code>head(variants)</code>.</p>
</div>
<!-- ### SOLUTION -->
<!-- ```{r} -->
<!-- variants <- read.csv("data/combined_tidy_vcf.csv") -->
<!-- ``` -->
<p>You can further investigate the structure of the data frame using the <code>str()</code> function:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">str</span>(variants)</a></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    801 obs. of  29 variables:
##  $ sample_id    : Factor w/ 3 levels &quot;SRR2584863&quot;,&quot;SRR2584866&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ CHROM        : Factor w/ 1 level &quot;CP000819.1&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##  $ POS          : int  9972 263235 281923 433359 473901 648692 1331794 1733343 2103887 2333538 ...
##  $ ID           : logi  NA NA NA NA NA NA ...
##  $ REF          : Factor w/ 59 levels &quot;A&quot;,&quot;ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG&quot;,..: 49 33 33 30 24 16 16 33 2 12 ...
##  $ ALT          : Factor w/ 57 levels &quot;A&quot;,&quot;AC&quot;,&quot;ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG&quot;,..: 31 46 46 29 25 46 1 1 4 15 ...
##  $ QUAL         : num  91 85 217 64 228 210 178 225 56 167 ...
##  $ FILTER       : logi  NA NA NA NA NA NA ...
##  $ INDEL        : logi  FALSE FALSE FALSE TRUE TRUE FALSE ...
##  $ IDV          : int  NA NA NA 12 9 NA NA NA 2 7 ...
##  $ IMF          : num  NA NA NA 1 0.9 ...
##  $ DP           : int  4 6 10 12 10 10 8 11 3 7 ...
##  $ VDB          : num  0.0257 0.0961 0.7741 0.4777 0.6595 ...
##  $ RPB          : num  NA 1 NA NA NA NA NA NA NA NA ...
##  $ MQB          : num  NA 1 NA NA NA NA NA NA NA NA ...
##  $ BQB          : num  NA 1 NA NA NA NA NA NA NA NA ...
##  $ MQSB         : num  NA NA 0.975 1 0.916 ...
##  $ SGB          : num  -0.556 -0.591 -0.662 -0.676 -0.662 ...
##  $ MQ0F         : num  0 0.167 0 0 0 ...
##  $ ICB          : logi  NA NA NA NA NA NA ...
##  $ HOB          : logi  NA NA NA NA NA NA ...
##  $ AC           : int  1 1 1 1 1 1 1 1 1 1 ...
##  $ AN           : int  1 1 1 1 1 1 1 1 1 1 ...
##  $ DP4          : Factor w/ 217 levels &quot;0,0,0,2&quot;,&quot;0,0,0,3&quot;,..: 3 132 73 141 176 104 61 74 133 137 ...
##  $ MQ           : int  60 33 60 60 60 60 60 60 60 60 ...
##  $ Indiv        : Factor w/ 3 levels &quot;/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ gt_PL        : Factor w/ 206 levels &quot;100,0&quot;,&quot;103,0&quot;,..: 16 10 134 198 142 127 93 142 9 80 ...
##  $ gt_GT        : int  1 1 1 1 1 1 1 1 1 1 ...
##  $ gt_GT_alleles: Factor w/ 57 levels &quot;A&quot;,&quot;AC&quot;,&quot;ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG&quot;,..: 31 46 46 29 25 46 1 1 4 15 ...</code></pre>
<div class="box">
<h4 id="exercise-22" class="unnumbered">EXERCISE</h4>
<p>Take a few minutes to familiarize yourself with this dataset. There are a lot of variables (how many?), so only worry about the ones listed below.</p>
<table>
<colgroup>
<col width="9%" />
<col width="90%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Column</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">sample_id</td>
<td align="left">sample ID</td>
</tr>
<tr class="even">
<td align="left">CHROM</td>
<td align="left">contig location where the variation occurs</td>
</tr>
<tr class="odd">
<td align="left">POS</td>
<td align="left">position within the contig where the variation occurs</td>
</tr>
<tr class="even">
<td align="left">REF</td>
<td align="left">reference genotype (forward strand)</td>
</tr>
<tr class="odd">
<td align="left">ALT</td>
<td align="left">sample genotype (forward strand)</td>
</tr>
<tr class="even">
<td align="left">QUAL</td>
<td align="left">Phred-scaled probablity that the observed variant exists at this site (higher is better)</td>
</tr>
<tr class="odd">
<td align="left">INDEL</td>
<td align="left">whether the variant is an indel</td>
</tr>
<tr class="even">
<td align="left">IDV</td>
<td align="left">length of indel</td>
</tr>
<tr class="odd">
<td align="left">IMF</td>
<td align="left">maximum fraction of reads supporting an indel</td>
</tr>
<tr class="even">
<td align="left">DP</td>
<td align="left">the depth per allele by sample and coverage</td>
</tr>
<tr class="odd">
<td align="left">MQ</td>
<td align="left">mapping quality</td>
</tr>
<tr class="even">
<td align="left">Indiv</td>
<td align="left">name of file</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="plotting-with-ggplot2" class="section level2">
<h2><span class="header-section-number">9.2</span> Plotting with ggplot2</h2>
<p>We start by loading the <code>ggplot2</code> package:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">library</span>(ggplot2)</a></code></pre></div>
<p><code>ggplot2</code> is a plotting package that makes it simple to create complex plots from data in a data frame. It provides a more programmatic interface for specifying what variables to plot, how they are displayed, and general visual properties. Therefore, we only need minimal changes if the underlying data change or if we decide to change from a bar plot to a scatter plot. This helps in creating publication quality plots with minimal amounts of adjustments and tweaking.</p>
<p><code>ggplot2</code> functions like data in the ‘long’ format, i.e., a column for every dimension, and a row for every observation. Well-structured data will save you lots of time when making figures with <code>ggplot2</code>.</p>
<p>Graphics are built step by step by adding new elements. Adding layers in this fashion allows for extensive flexibility and customization of plots.</p>
<p>To build a ggplot, we will use the following basic template that can be used for different types of plots:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> <span class="op">&lt;</span>DATA<span class="op">&gt;</span>, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="op">&lt;</span>MAPPINGS<span class="op">&gt;</span>)) <span class="op">+</span><span class="st">  </span><span class="er">&lt;</span>GEOM_FUNCTION<span class="op">&gt;</span>()</a></code></pre></div>
<ol style="list-style-type: decimal">
<li>Specify which data set to use for the plot using the <code>data</code> argument:</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants)</a></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Define a “mapping” (using the aesthetic (<code>aes</code>) function), by selecting the variables to be plotted and specifying how to present them in the graph, e.g. as x/y positions or characteristics such as size, shape, color, etc:</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> POS, <span class="dt">y =</span> DP))</a></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Add “geoms” – graphical representations of the data in the plot (points, lines, bars). <code>ggplot2</code> offers many different geoms; we will use some common ones today, including:</li>
</ol>
<ul>
<li><code>geom_point()</code> for scatter plots, dot plots, etc.</li>
<li><code>geom_boxplot()</code> for boxplots.</li>
<li><code>geom_histogram()</code> for histograms.</li>
<li><code>geom_barplot()</code> for barplots.</li>
<li><code>geom_line()</code> for trend lines, time series, etc.</li>
</ul>
<p>To add a geom to the plot use the <code>+</code> operator. Because we have two continuous variables, let’s use <code>geom_point()</code> first:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> POS, <span class="dt">y =</span> DP)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>()</a></code></pre></div>
<p><img src="IntroRWorkshop_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>The <code>+</code> in the <code>ggplot2</code> package is particularly useful because it allows you to modify existing <code>ggplot</code> objects. This means you can easily set up plot templates and conveniently explore different types of plots, so the above plot can also be generated with code like this:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co"># Assign plot to a variable</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2">coverage_plot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> POS, <span class="dt">y =</span> DP))</a>
<a class="sourceLine" id="cb8-3" data-line-number="3"></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="co"># Draw the plot</span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5">coverage_plot <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-6" data-line-number="6"><span class="st">    </span><span class="kw">geom_point</span>()</a></code></pre></div>
<p>Notes:</p>
<ul>
<li>Anything you put in the <code>ggplot()</code> function can be seen by any geom layers that you add (i.e., these are universal plot settings). This includes the x- and y-axis mapping you set up in <code>aes()</code>.</li>
<li>You can also specify mappings for a given geom independently of the mappings defined globally in the <code>ggplot()</code> function.</li>
<li>The <code>+</code> sign used to add new layers must be placed at the end of the line containing the previous layer. If, instead, the <code>+</code> sign is added at the beginning of the line containing the new layer, <code>ggplot2</code> will not add the new layer and will return an error message.</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co"># This is the correct syntax for adding layers</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2">coverage_plot <span class="op">+</span></a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_point</span>()</a>
<a class="sourceLine" id="cb9-4" data-line-number="4"></a>
<a class="sourceLine" id="cb9-5" data-line-number="5"><span class="co"># This will not add the new layer and will return an error message</span></a>
<a class="sourceLine" id="cb9-6" data-line-number="6">coverage_plot</a>
<a class="sourceLine" id="cb9-7" data-line-number="7">  <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</a></code></pre></div>
<div id="principles-of-effective-display" class="section level3 unnumbered">
<h3>Principles of effective display</h3>
<p>SOURCE: (Whitlock &amp; Schluter, The Analysis of Biological Data)[<a href="http://whitlockschluter.zoology.ubc.ca/" class="uri">http://whitlockschluter.zoology.ubc.ca/</a>]</p>
<p>We will follow these metrics to create and evaluate figures:<br />
1. Show the data<br />
2. Make patterns in the data easy to see<br />
3. Represent magnitudes honestly<br />
4. Draw graphical elements clearly, minimizing clutter</p>
<div class="box">
<h4 id="exercise-23" class="unnumbered">EXERCISE</h4>
<p>Create a scatter plot (using the <code>geom_point()</code> function for quality (<code>QUAL</code>) versus coverage depth (<code>DP</code>).</p>
</div>
<!-- ### SOLUTION -->
<!-- ```{r} -->
<!-- ggplot(data = variants, aes(x = QUAL, y = DP)) +  -->
<!--   geom_point() -->
<!-- ``` -->
</div>
<div id="building-plots-iteratively" class="section level3 unnumbered">
<h3>Building plots iteratively</h3>
<p>Building plots with <code>ggplot2</code> is typically an iterative process. We start by defining the dataset we’ll use, lay out the axes, and choose a geom:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> POS, <span class="dt">y =</span> DP)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>()</a></code></pre></div>
<p><img src="IntroRWorkshop_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p>Then, we start modifying this plot to extract more information from it. For instance, we can add transparency (<code>alpha</code>) to avoid overplotting:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> POS, <span class="dt">y =</span> DP)) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>)</a></code></pre></div>
<p><img src="IntroRWorkshop_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<p>We can also add colors for all the points:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> POS, <span class="dt">y =</span> DP)) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>, <span class="dt">color =</span> <span class="st">&quot;blue&quot;</span>)</a></code></pre></div>
<p><img src="IntroRWorkshop_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>Or to color each species in the plot differently, you could use a vector as an input to the argument color. <code>ggplot2</code> will provide a different color corresponding to different values in the vector. Here is an example where we color with <code>sample_id</code>:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> POS, <span class="dt">y =</span> DP, <span class="dt">color =</span> sample_id)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>)</a></code></pre></div>
<p><img src="IntroRWorkshop_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>Notice that we can change the geom layer and colors will be still determined by <code>sample_id</code>:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> POS, <span class="dt">y =</span> DP, <span class="dt">color =</span> sample_id)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>)</a></code></pre></div>
<p><img src="IntroRWorkshop_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<p>To make our plot more readable, we can add axis labels:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> POS, <span class="dt">y =</span> DP, <span class="dt">color =</span> sample_id)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb15-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb15-3" data-line-number="3"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Base Pair Position&quot;</span>, </a>
<a class="sourceLine" id="cb15-4" data-line-number="4">       <span class="dt">y =</span> <span class="st">&quot;Read Depth (DP)&quot;</span>)</a></code></pre></div>
<p><img src="IntroRWorkshop_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<div class="box">
<h4 id="exercise-24" class="unnumbered">EXERCISE</h4>
<p>Create a scatter plot of mapping quality (<code>MQ</code>) over position (<code>POS</code>) with the samples showing in different colors. Make sure to give your plot relevant axis labels.</p>
</div>
<!-- Solution -->
<!-- ```{r} -->
<!-- ggplot(data = variants, aes(x = POS, y = MQ, color = sample_id)) + -->
<!--   geom_point() + -->
<!--   labs(x = "Base Pair Position", -->
<!--        y = "Mapping Quality (MQ)") -->
<!-- ``` -->
</div>
</div>
<div id="faceting" class="section level2">
<h2><span class="header-section-number">9.3</span> Faceting</h2>
<p><code>ggplot2</code> has a special technique called faceting that allows the user to split one plot into multiple plots based on a factor included in the dataset. We will use it to split our mapping quality plot into three panels, one for each sample.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> POS, <span class="dt">y =</span> MQ, <span class="dt">color =</span> sample_id)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2"><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb16-3" data-line-number="3"><span class="st"> </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Base Pair Position&quot;</span>, </a>
<a class="sourceLine" id="cb16-4" data-line-number="4">      <span class="dt">y =</span> <span class="st">&quot;Mapping Quality (MQ)&quot;</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb16-5" data-line-number="5"><span class="st"> </span><span class="kw">facet_grid</span>(. <span class="op">~</span><span class="st"> </span>sample_id)</a></code></pre></div>
<p><img src="IntroRWorkshop_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<p>This looks ok, but it would be easier to read if the plot facets were stacked vertically rather than horizontally. The <code>facet_grid</code> geometry allows you to explicitly specify how you want your plots to be arranged via formula notation (<code>rows ~ columns</code>); a <code>.</code> can be used as a placeholder that indicates only one row or column).</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> POS, <span class="dt">y =</span> MQ, <span class="dt">color =</span> sample_id)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb17-3" data-line-number="3"><span class="st"> </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Base Pair Position&quot;</span>, </a>
<a class="sourceLine" id="cb17-4" data-line-number="4">      <span class="dt">y =</span> <span class="st">&quot;Mapping Quality (MQ)&quot;</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb17-5" data-line-number="5"><span class="st"> </span><span class="kw">facet_grid</span>(sample_id <span class="op">~</span><span class="st"> </span>.)</a></code></pre></div>
<p><img src="IntroRWorkshop_files/figure-html/unnamed-chunk-17-1.png" width="672" /></p>
<p>Usually plots with white background look more readable when printed. We can set the background to white using the function <code>theme_bw()</code>. Additionally, you can remove the grid:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> POS, <span class="dt">y =</span> MQ, <span class="dt">color =</span> sample_id)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-3" data-line-number="3"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Base Pair Position&quot;</span>, </a>
<a class="sourceLine" id="cb18-4" data-line-number="4">       <span class="dt">y =</span> <span class="st">&quot;Mapping Quality (MQ)&quot;</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-5" data-line-number="5"><span class="st">  </span><span class="kw">facet_grid</span>(sample_id <span class="op">~</span><span class="st"> </span>.) <span class="op">+</span></a>
<a class="sourceLine" id="cb18-6" data-line-number="6"><span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb18-7" data-line-number="7"><span class="st">  </span><span class="kw">theme</span>(<span class="dt">panel.grid =</span> <span class="kw">element_blank</span>())</a></code></pre></div>
<p><img src="IntroRWorkshop_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<div class="box">
<h4 id="exercise-25" class="unnumbered">EXERCISE</h4>
<p>Use what you just learned to create a scatter plot of PHRED scaled quality (<code>QUAL</code>) over position (<code>POS</code>) with the samples showing in different colors. Make sure to give your plot relevant axis labels.</p>
</div>
<!-- ### SOLUTION -->
<!-- ```{r} -->
<!-- ggplot(data = variants, aes(x = POS, y = QUAL, color = sample_id)) + -->
<!--   geom_point() + -->
<!--   labs(x = "Base Pair Position", -->
<!--        y = "PHRED-sacled Quality (QUAL)") + -->
<!--   facet_grid(sample_id ~ .) -->
<!-- ``` -->
</div>
<div id="barplots" class="section level2">
<h2><span class="header-section-number">9.4</span> Barplots</h2>
<p>We can create barplots using the <code>geom_bar</code> geom. Let’s make a barplot showing the number of variants for each sample that are indels.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> INDEL, <span class="dt">fill =</span> sample_id)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_bar</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb19-3" data-line-number="3"><span class="st">  </span><span class="kw">facet_grid</span>(sample_id <span class="op">~</span><span class="st"> </span>.)</a></code></pre></div>
<p><img src="IntroRWorkshop_files/figure-html/unnamed-chunk-19-1.png" width="672" /></p>
<div class="box">
<h4 id="exercise-26" class="unnumbered">EXERCISE</h4>
<p>Since we already have the sample_id labels on the individual plot facets, we don’t need the legend. Use the help file for <code>geom_bar</code> and any other online resources you want to use to remove the legend from the plot.</p>
</div>
<!-- ### SOLUTION -->
<!-- ```{r} -->
<!-- ggplot(data = variants, aes(x = INDEL, color = sample_id)) +  -->
<!--    geom_bar(show.legend = F) + -->
<!--    facet_grid(sample_id ~ .) -->
<!-- ``` -->
</div>
<div id="histograms" class="section level2">
<h2><span class="header-section-number">9.5</span> Histograms</h2>
<p>Sometimes it can be useful to plot a single variable at a time. Usually this is for exploratory purposes - to get a feel for a variable. To do this, use the <code>geom_histogram()</code> function. For example, to look at the distribution of Qualities:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(QUAL)) <span class="op">+</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_histogram</span>()</a></code></pre></div>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="IntroRWorkshop_files/figure-html/unnamed-chunk-20-1.png" width="672" /></p>
<p>R is giving us a warning. We can choose to ignore this if we are able to see what we want in the plot. Otherwise, we can change the <code>binwidth</code> by using the binwidth argument:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(QUAL)) <span class="op">+</span></a>
<a class="sourceLine" id="cb22-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">binwidth =</span> <span class="dv">5</span>)</a></code></pre></div>
<img src="IntroRWorkshop_files/figure-html/unnamed-chunk-21-1.png" width="672" />
<div class="box">
<h4 id="exercise-27" class="unnumbered">EXERCISE</h4>
<p>Create a plot that shows the distribution of Read Depth (DP) for each sample separately.</p>
<p><b>Bonus challenge.</b> Create this plot for only two of the three samples IDs - the two with many fewer variants.</p>
</div>
<!-- ### Solution -->
<!-- ```{r} -->
<!-- ggplot(data = variants, aes(DP)) + -->
<!--   geom_histogram(binwidth = 2) + -->
<!--   facet_grid(sample_id ~ .) -->
<!-- ``` -->
<!-- ### Bonus solution -->
<!-- ```{r} -->
<!-- # subset the data -->
<!-- two <- filter(variants, sample_id != "SRR2584866") -->
<!-- ggplot(data = two, aes(DP)) + -->
<!--   geom_histogram(binwidth = 2) + -->
<!--   facet_grid(sample_id ~ .) -->
<!-- ``` -->
</div>
<div id="boxplots" class="section level2">
<h2><span class="header-section-number">9.6</span> Boxplots</h2>
<p>When there are a large number of values for a certain variable (like for one of the samples above), boxplots can be a useful way to display summary statistics like the median, and “spread” of a variable.</p>
<div id="the-five-number-summary" class="section level3 unnumbered">
<h3>The Five Number Summary</h3>
<p>The five number summary gives a quick look at the features of numerical variables. It consists of the variables:</p>
<ul>
<li>minimum</li>
<li>1st quartile</li>
<li>median</li>
<li>3rd quartile</li>
<li>maximum</li>
</ul>
<p><strong>QUANTILES:</strong> The <em>pth</em> percentile of a data set sorted from smallest to largest is the value such that <em>p</em> percent of the data are at or below this value. The quartiles are special percentiles; the 1st quartile is the 25th percentile, and the 3rd quartile is the 75th percentile. The median is also a quartile – it is the 50th percentile.</p>
<p>Within these five numbers is a lot of useful data!</p>
<ul>
<li>the median gives a measure of the center of the data</li>
<li>the minimum and maximum give the range of the data</li>
<li>the 1st and 3rd quartiles give a sense of the spread of the data, especially when compared to the minimum, maximum, and median</li>
</ul>
<p>To create a boxplot, use the <code>geom_boxplot()</code> function:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="kw">ggplot</span>(<span class="dt">data =</span> variants, <span class="kw">aes</span>(<span class="dt">x =</span> sample_id, <span class="dt">y =</span> DP)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_boxplot</span>()</a></code></pre></div>
<p><img src="IntroRWorkshop_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
<div class="box">
<h4 id="exercise-28" class="unnumbered">EXERCISE</h4>
<ol style="list-style-type: decimal">
<li>Change the colour of each box in the above plot to match the sample_id colours in the plots above.</li>
<li>Log transform the y-axis.</li>
</ol>
<p><b>Bonus challenge:</b> Change the name of the legend to “Sample ID”.</p>
</div>
<!-- ### SOLUTION -->
<!-- ```{r} -->
<!-- ggplot(data = variants, aes(x = sample_id, y = log(DP), color = sample_id)) + -->
<!--   geom_boxplot() -->
<!-- ``` -->
</div>
</div>
<div id="themes" class="section level2">
<h2><span class="header-section-number">9.7</span> Themes</h2>
<p>In addition to <code>theme_bw()</code>, which changes the plot background to white, <code>ggplot2</code> comes with several other themes which can be useful to quickly change the look of your visualization. The complete list of themes is available at <a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="uri">https://ggplot2.tidyverse.org/reference/ggtheme.html</a>. <code>theme_minimal()</code> and <code>theme_light()</code> are popular, and <code>theme_void()</code> can be useful as a starting point to create a new hand-crafted theme.</p>
<p>The <code>ggthemes</code> package provides a wide variety of options (including an Excel 2003 theme). The <code>ggplot2</code> <a href="https://www.ggplot2-exts.org">extensions website</a> provides a list of packages that extend the capabilities of <code>ggplot2</code>, including additional themes.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="manipulating-data-in-the-tidyverse.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="project.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["IntroRWorkshop.pdf", "IntroRWorkshop.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
